### ISSUE TITLE
New rule for str.join usage simplification

### ISSUE THREAD

**Issue by victorsamun** [astral-sh/ruff#15468]
Method `str.join` takes an iterable, so we don't need create list or tuple explicitly.

For example, `", ".join(list(data))` and `", ".join(tuple(data))` simplifies to `", ".join(data)`.
This can be done automatically.

**Comment by dylwil3** [comment]
A related thing we could do is extend [`C419`](https://docs.astral.sh/ruff/rules/unnecessary-comprehension-in-call/) to include this string method, and maybe other builtins and standard library methods that accept iterables. That would cover some instances of the proposed lint.

**Comment by nijel** [comment]
Unnecessary list for `str.join` is indeed often used in comprehension:

```py
        return b"\r\n".join(
            [
                key.encode("ascii") + b": " + value.encode("latin-1")
                for key, value in self.headers.items()
            ]
        )
```

So C419 might be the good place to land this.

**Comment by nijel** [comment]
Hmm, it turns out to be actually a bad idea. Passing list to `str.join` is more effective than iterator because it does two passes over the sequence. Also see https://stackoverflow.com/a/9061024. I've verified the numbers in Python 3.13 and 3.14.

**Comment by MichaReiser** [comment]
Related https://github.com/astral-sh/ruff/issues/12912 and https://github.com/astral-sh/ruff/issues/10838

**Comment by nijel** [comment]
I've already spent a lot in this rabbit hole to say it's a bit different. With `sum`/`min`/`max`/`all`/`any` there is at least something to gain (less memory allocations). For `join` there is nothing to gain as it internally converts iterator to a list as the first step (because it needs two passes on the data).

**Comment by user27182** [comment]
It seems this is already being caught by C400 and C416? e.g. with
``` py
'-'.join([s for s in 'abc'])  # C400 error
'-'.join(list(s for s in 'abc'))  # C416 error
'-'.join(s for s in 'abc')  # OK
```
https://play.ruff.rs/a59beb82-bdf3-46a8-a838-9795618a0d44

But it seems like this is a (separate?) bug? Because with the first line, the diagnostic suggests rewriting using `list()`, which is what I did for the second line. But the diagnostic for the second line also suggests using `list()`.

The suggestion to use `list()` is confusing. Both should instead suggest using `join()` (or perhaps having a list is better performance-wise as suggested earlier https://github.com/astral-sh/ruff/issues/15468#issuecomment-2711522837).

Should I open a separate issue for this misleading diagnostic message?

**Comment by ntBre** [comment]
If you apply the fix for `C416` on line 1, you'll see that it produces:

```py
'-'.join(list('abc'))
```

not the `list` containing a generator like in your second line. I'm not sure the diagnostic message could be improved without becoming much longer, but if you have a suggestion feel free to open an issue! I do think it helps a lot to see the diagnostic with the fix, which is still in preview:

```
❯ ruff check try.py --select C416 --unsafe-fixes --preview
C416 [*] Unnecessary list comprehension (rewrite using `list()`)
 --> try.py:1:10
  |
1 | '-'.join([s for s in 'abc'])
  |          ^^^^^^^^^^^^^^^^^^
2 | '-'.join(list(s for s in 'abc'))
3 | '-'.join(s for s in 'abc')
  |
help: Rewrite using `list()`
  - '-'.join([s for s in 'abc'])
1 + '-'.join(list('abc'))
2 | '-'.join(list(s for s in 'abc'))
3 | '-'.join(s for s in 'abc')
4 |
note: This is an unsafe fix and may change runtime behavior
```

My understanding is that the rule suggested here would then simplify `'-'.join(list('abc'))` further to `'-'.join('abc')`.

**Comment by user27182** [comment]
Thanks for the clarification! Indeed there is no (additional) issue then. Seems like I had a bad example to start with.
